source/common/router/router.cc

```diff
@@ -557,7 +557,7 @@ void Filter::onUpstreamReset(UpstreamResetType type,
     }
 
     RetryStatus retry_status =
-        retry_state_->shouldRetry(nullptr, reset_reason, [this]() -> void { doRetry(); });
+        retry_state_->shouldRetryReset(reset_reason.value(), [this]() -> void { doRetry(); });
```

Comment:
How are we guaranteed that this option as a `value()`? It's not clear to me from the code flow. Can you add a comment or potentially look at changing the logic if always has a value?

Also, while you are here, can you change the call signature to pass the `StreamResetReason` by value and not reference? It will be faster (my own badly written code most likely).

Reply:
yeah I think I don't need an optional anymore, will do that

Reply:
nvm there is a place where it might not have a value (`onResponseTimeout()`). I actually take care of that case and remove the optional in https://github.com/envoyproxy/envoy/pull/6142. For this PR i'll put a comment explaining why it will always have a value